bothan-core/src/registry/processor/weighted_median.rs

```diff
@@ -0,0 +1,217 @@
+        let serialized_decimal_map: HashMap<String, [u8; 16]> =
+            BorrowDecode::borrow_decode(decoder)?;
+        let source_weights = serialized_decimal_map
+            .into_iter()
+            .map(|(k, v)| (k, Decimal::deserialize(v)))
+            .collect();
+
+        let minimum_cumulative_weight = Decimal::deserialize(BorrowDecode::borrow_decode(decoder)?);
+
+        Ok(WeightedMedianProcessor {
+            source_weights,
+            minimum_cumulative_weight,
+        })
+    }
+}
+
+impl WeightedMedianProcessor {
+    /// Creates a new `WeightedMedianProcessor`.
+    pub fn new(
+        source_weights: HashMap<String, Decimal>,
+        minimum_cumulative_weight: Decimal,
+    ) -> Self {
+        WeightedMedianProcessor {
+            source_weights,
+            minimum_cumulative_weight,
+        }
+    }
+}
+
+impl Process<(String, Decimal), Decimal> for WeightedMedianProcessor {
+    /// Processes the given data and returns the weighted median. If the cumulative weights of the
+    /// data sources are less than the minimum cumulative weight or the source associated
+    /// with the data does not have an assigned weight, it returns an error.
+    fn process(&self, data: Vec<(String, Decimal)>) -> Result<Decimal, ProcessError> {
+        let cumulative_weight = data.iter().try_fold(Decimal::zero(), |acc, (source, _)| {
+            self.source_weights
+                .get(source)
+                .map(|weight| acc + weight)
+                .ok_or(ProcessError::new(format!("Unknown source {source}")))
+        })?;
+
+        if cumulative_weight < self.minimum_cumulative_weight {
+            return Err(ProcessError::new(
+                "Not enough sources to calculate weighted median",
+            ));
+        }
+
+        let values = data
+            .into_iter()
+            .map(|(source, value)| {
+                self.source_weights
+                    .get(&source)
+                    .map(|weight| (value, *weight))
+                    .ok_or(ProcessError::new(format!("Unknown source {source}")))
+            })
+            .collect::<Result<Vec<(Decimal, Decimal)>, ProcessError>>()?;
+
+        let median = compute_weighted_median(values)
+            .ok_or_else(|| ProcessError::new("Failed to calculate weighted median"))?;
+        Ok(median)
+    }
+}
+
+fn compute_weighted_median<T>(mut values: Vec<(T, Decimal)>) -> Option<T>
+where
+    T: Ord + Add<Output = T> + Div<Output = T> + FromPrimitive,
+{
+    values.sort_by(|(v1, _), (v2, _)| v1.cmp(v2));
+
+    let sum = values
+        .iter()
+        .fold(Decimal::zero(), |acc, (_, weight)| acc + weight);
+
+    let mid = sum / Decimal::new(2, 0);
+
+    let mut cumulative_weight = Decimal::zero();
+    let mut iter = values.into_iter();
+    while let Some((value, weight)) = iter.next() {
+        cumulative_weight += weight;
+        match cumulative_weight.cmp(&mid) {
```

Comment:
instead of using mid. i think it's more elegant to compare `cumulative_weight * 2` vs `sum` so we don't need to worry about division precision.